universal_transfer_operator/src/universal_transfer_operator/data_providers/filesystem/sftp.py

```diff
@@ -0,0 +1,143 @@
+from __future__ import annotations
+
+from functools import cached_property
+from urllib.parse import urlparse, urlunparse
+
+import attr
+import smart_open
+from airflow.providers.sftp.hooks.sftp import SFTPHook
+
+from universal_transfer_operator.constants import Location, TransferMode
+from universal_transfer_operator.data_providers.filesystem.base import (
+    BaseFilesystemProviders,
+    FileStream,
+)
+from universal_transfer_operator.datasets.file.base import File
+from universal_transfer_operator.utils import TransferParameters
+
+
+class SFTPDataProvider(BaseFilesystemProviders):
+    """
+    DataProviders interactions with GS Dataset.
+    """
+
+    def __init__(
+        self,
+        dataset: File,
+        transfer_params: TransferParameters = attr.field(
+            factory=TransferParameters,
+            converter=lambda val: TransferParameters(**val) if isinstance(val, dict) else val,
+        ),
+        transfer_mode: TransferMode = TransferMode.NONNATIVE,
+    ):
+        super().__init__(
+            dataset=dataset,
+            transfer_params=transfer_params,
+            transfer_mode=transfer_mode,
+        )
+        self.transfer_mapping = {
+            Location.S3,
+            Location.GS,
+        }
+
+    @cached_property
+    def hook(self) -> SFTPHook:
+        """Return an instance of the SFTPHook Airflow hook."""
+        return SFTPHook(ssh_conn_id=self.dataset.conn_id)
+
+    @property
+    def paths(self) -> list[str]:
+        """Resolve SFTP file paths with prefix"""
+        url = urlparse(self.dataset.path)
+        uri = self.get_uri()
+        full_paths = []
+        prefixes = self.hook.get_tree_map(url.netloc, prefix=url.netloc + url.path)
+        for keys in prefixes:
+            if len(keys) > 0:
+                full_paths.extend(keys)
+        # paths = ["/" + path for path in full_paths]
+        paths = [uri + "/" + path for path in full_paths]
+        return paths
+
+    @property
+    def transport_params(self) -> dict:
+        """get SFTP credentials for storage"""
+        client = self.hook.get_connection(self.dataset.conn_id)
+        extra_options = client.extra_dejson
+        if "key_file" in extra_options:
+            key_file = extra_options.get("key_file")
+            return {"connect_kwargs": {"key_filename": key_file}}
+        elif client.password:
+            return {"connect_kwargs": {"password": client.password}}
+        raise ValueError("SFTP credentials are not set in the connection.")
+
+    def get_uri(self):
+        client = self.hook.get_connection(self.dataset.conn_id)
+        return client.get_uri()
+
+    @staticmethod
+    def get_url_netloc(complete_url, dst_url):
```

Comment:
Please add docstring and types

Reply:
Removed method since it was no longer needed.